kv/kvclient/rangefeed: TestRangefeedWithLabelsOption failed

kv/kvclient/rangefeed.TestRangefeedWithLabelsOption [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/13135765?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/13135765?buildTab=artifacts#/) on master @ [cbcdca3b6a5e54ee936efe17c854a73bbbab74af](https://github.com/cockroachdb/cockroach/commits/cbcdca3b6a5e54ee936efe17c854a73bbbab74af):

Fatal error:

```
panic: use of Span after Finish. Span: /cockroach.roachpb.Internal/RangeFeed. Finish previously called at: <stack not captured. Set debugUseAfterFinish> [recovered]
	panic: use of Span after Finish. Span: /cockroach.roachpb.Internal/RangeFeed. Finish previously called at: <stack not captured. Set debugUseAfterFinish>

```
Stack: 

```
goroutine 101783 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0x5e5d3e8?, {0xac619d8, 0xc003c86330})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:230 +0x74
panic({0x7a097e0?, 0xc005581df0?})
	GOROOT/src/runtime/panic.go:920 +0x270
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).detectUseAfterFinish(0xc008d66800)
	github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:182 +0x274
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).RecordingType(0xc008d66800)
	github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:442 +0x27
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).IsVerbose(...)
	github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:450
github.com/cockroachdb/cockroach/pkg/util/log.getSpanOrEventLog({0xac619d8, 0xc003c86030})
	github.com/cockroachdb/cockroach/pkg/util/log/trace.go:92 +0x45
github.com/cockroachdb/cockroach/pkg/util/log.HasSpanOrEvent(...)
	github.com/cockroachdb/cockroach/pkg/util/log/trace.go:274
github.com/cockroachdb/cockroach/pkg/storage.pebbleLogger.IsTracingEnabled(...)
	github.com/cockroachdb/cockroach/pkg/storage/pebble.go:759
github.com/cockroachdb/pebble/sstable.(*Reader).readBlock(0xc000baee00, {0xac619d8, 0xc003c86030}, {0x0?, 0x1e?}, 0x0, {0xac68bd0, 0xc0097b8c28}, 0xc007b303c0, 0xc0097b8cf0, ...)
	github.com/cockroachdb/pebble/sstable/external/com_github_cockroachdb_pebble/sstable/reader.go:598 +0x674
github.com/cockroachdb/pebble/sstable.(*singleLevelIterator).loadBlock(0xc0097b8900, 0x33?)
	github.com/cockroachdb/pebble/sstable/external/com_github_cockroachdb_pebble/sstable/reader_iter_single_lvl.go:414 +0x8a5
github.com/cockroachdb/pebble/sstable.(*singleLevelIterator).seekGEHelper(0xc0097b8900, {0xc006ac6533, 0xd, 0xd}, 0x0, 0x0)
	github.com/cockroachdb/pebble/sstable/external/com_github_cockroachdb_pebble/sstable/reader_iter_single_lvl.go:729 +0x785
github.com/cockroachdb/pebble/sstable.(*singleLevelIterator).SeekGE(0xc0097b8900, {0xc006ac6533, 0xd, 0xd}, 0x0)
	github.com/cockroachdb/pebble/sstable/external/com_github_cockroachdb_pebble/sstable/reader_iter_single_lvl.go:639 +0x315
github.com/cockroachdb/pebble.(*levelIter).SeekGE(0xc007b30cd8, {0xc006ac6533, 0xd, 0xd}, 0x0)
	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/level_iter.go:745 +0x1df
github.com/cockroachdb/pebble.(*mergingIter).seekGE(0xc007b305f0, {0xc006ac6533, 0xd, 0xd}, 0x0, 0x0)
	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/merging_iter.go:1051 +0x743
github.com/cockroachdb/pebble.(*mergingIter).SeekGE(0xc007b305f0, {0xc006ac6533, 0xd, 0xd}, 0xd?)
	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/merging_iter.go:1113 +0xd9
github.com/cockroachdb/pebble.(*Iterator).SeekGEWithLimit(0xc007b30000, {0xc006ac6533, 0xd, 0xd}, {0x0, 0x0, 0x0})
	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/iterator.go:1319 +0x106a
github.com/cockroachdb/pebble.(*Iterator).SeekGE(...)
	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/iterator.go:1203
github.com/cockroachdb/cockroach/pkg/storage/pebbleiter.(*assertionIter).SeekGE(0xc0071403c0, {0xc006ac6533, 0xd, 0xd})
	github.com/cockroachdb/cockroach/bazel-out/k8-fastbuild/bin/pkg/storage/pebbleiter/gen-crdb_test_on.go:171 +0xf5
github.com/cockroachdb/cockroach/pkg/storage.(*pebbleIterator).SeekEngineKeyGE(0xc005716e00, {{0xc006ac6524, 0xc, 0xc}, {0x0, 0x0, 0x0}})
	github.com/cockroachdb/cockroach/pkg/storage/pebble_iterator.go:389 +0x21c
github.com/cockroachdb/cockroach/pkg/storage.(*LockTableIterator).SeekEngineKeyGE(0xc0052e3d00, {{0xc006ac6524, 0xc, 0xc}, {0x0, 0x0, 0x0}})
	github.com/cockroachdb/cockroach/pkg/storage/lock_table_iterator.go:154 +0x1ac
github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed.(*SeparatedIntentScanner).ConsumeIntents(0xc003908368, {0x0?, 0x0?}, {0xc003ac0be8, 0x5, 0x8}, {0x7f247924e300?, 0xc000d348c8?, 0x0?}, 0xc0026dda00)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed/task.go:129 +0x1bd
github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed.(*initResolvedTSScan).iterateAndConsume(0xc00762b2c0, {0xac619d8?, 0xc003c86330})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed/task.go:75 +0x1dc
github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed.(*initResolvedTSScan).Run(0xc00762b2c0, {0xac619d8, 0xc003c86330})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed/task.go:62 +0xc5
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:484 +0x279
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 101782
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:475 +0x639
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestRangefeedWithLabelsOption
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/8a396ce5ffb24698ad44a501ceca79c0/logTestRangefeedWithLabelsOption2160701887
    test_log_scope.go:81: use -show-logs to present logs inline
=== RUN   TestRangefeedWithLabelsOption/feed_type=single/scheduler=false
    conditional_wrap.go:190: 
        pkg/kv/kvclient/rangefeed_test/pkg/kv/kvclient/rangefeed/rangefeed_external_test.go:1170: (TestRangefeedWithLabelsOption.func1)
        	NOTICE: .SplitRange() called via implicit interface StorageLayerInterface;
        HINT: consider using .StorageLayer().SplitRange() instead.
```
</p>
</details>

Parameters:
 - <code>TAGS=bazel,gss</code>
 - <code>stress=true</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/replication
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRangefeedWithLabelsOption.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-34628

--- a/pkg/kv/kvserver/rangefeed/task.go
+++ b/pkg/kv/kvserver/rangefeed/task.go
@@ -107,13 +107,19 @@ func NewSeparatedIntentScanner(
 ) (IntentScanner, error) {
 	lowerBound, _ := keys.LockTableSingleKey(span.Key.AsRawKey(), nil)
 	upperBound, _ := keys.LockTableSingleKey(span.EndKey.AsRawKey(), nil)
-	iter, err := storage.NewLockTableIterator(ctx, reader, storage.LockTableIteratorOptions{
-		LowerBound: lowerBound,
-		UpperBound: upperBound,
-		// Ignore Shared and Exclusive locks. We only care about intents.
-		MatchMinStr:  lock.Intent,
-		ReadCategory: storage.RangefeedReadCategory,
-	})
+	iter, err := storage.NewLockTableIterator(
+		// Do not use ctx, since it is not the ctx passed in when ConsumeIntents
+		// is called. See https://github.com/cockroachdb/cockroach/issues/116440.
+		//
+		// NB: the storage iterator does not respect context cancellation, and
+		// only uses it for tracing.
+		context.Background(), reader, storage.LockTableIteratorOptions{
+			LowerBound: lowerBound,
+			UpperBound: upperBound,
+			// Ignore Shared and Exclusive locks. We only care about intents.
+			MatchMinStr:  lock.Intent,
+			ReadCategory: storage.RangefeedReadCategory,
+		})
 	if err != nil {
 		return nil, err
 	}
@@ -126,6 +132,8 @@ func (s *SeparatedIntentScanner) ConsumeIntents(
 ) error {
 	ltStart, _ := keys.LockTableSingleKey(startKey, nil)
 	var meta enginepb.MVCCMetadata
+	// TODO(sumeer): ctx is not used for iteration. Fix by adding a method to
+	// EngineIterator to replace the context.
 	for valid, err := s.iter.SeekEngineKeyGE(storage.EngineKey{Key: ltStart}); ; valid, err = s.iter.NextEngineKey() {
 		if err != nil {
 			return err

